document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -564,6 +512,250 @@ public boolean bulkUpsert(Map<Key, Document> documents) {
+    PreparedStatement preparedStatement = client.prepareStatement(sqlQuery);
+    enrichPreparedStatementWithParams(preparedStatement, params);
+    return preparedStatement;
+  }
+
+  @VisibleForTesting
+  protected void enrichPreparedStatementWithParams(
+      PreparedStatement preparedStatement, Params params) throws RuntimeException {
+    params
+        .getObjectParams()
+        .forEach(
+            (k, v) -> {
+              try {
+                if (isValidType(v)) {
+                  preparedStatement.setObject(k, v);
+                } else {
+                  throw new UnsupportedOperationException("Un-supported object types in filter");
+                }
+              } catch (SQLException e) {
+                LOGGER.error("SQLException setting Param. key: {}, value: {}", k, v);
+              }
+            });
+  }
+
+  @Override
+  public void drop() {
+    String dropTableSQL = String.format("DROP TABLE IF EXISTS %s", collectionName);
+    try (PreparedStatement preparedStatement = client.prepareStatement(dropTableSQL)) {
+      preparedStatement.executeUpdate();
+    } catch (SQLException e) {
+      LOGGER.error("Exception deleting table name: {}", collectionName);
+    }
+  }
+
+  private BulkUpdateResult bulkRemoveOnArrayValue(BulkArrayValueUpdateRequest request)
+      throws IOException {
+    Map<Key, Document> upsertMap = new HashMap<>();
+    Map<String, String> idToTenantIdMap = getDocIdToTenantIdMap(request);
+    HashSet<JsonNode> subDocs = new HashSet<>();
+    for (Document subDoc : request.getSubDocuments()) {
+      subDocs.add(getDocAsJSON(subDoc));
+    }
+    Iterator<Document> docs = searchDocsForKeys(request.getKeys());
+    while (docs.hasNext()) {
+      JsonNode docRoot = getDocAsJSON(docs.next());
+      JsonNode currNode;
+      currNode = getJsonNodeAtPath(request.getSubDocPath(), docRoot, false);
+      if (currNode == null) {
+        LOGGER.warn(
+            "Removal path does not exist for {}, continuing with other documents",
+            docRoot.get(ID).textValue());
+        continue;
+      }
+      HashSet<JsonNode> existingItems = new HashSet<>();
+      ArrayNode candidateArray = (ArrayNode) currNode;
+      candidateArray.forEach(existingItems::add);
+      existingItems.removeAll(subDocs);
+      candidateArray.removeAll();
+      candidateArray.addAll(existingItems);
+      String id = docRoot.findValue(ID).asText();
+      upsertMap.put(new SingleValueKey(idToTenantIdMap.get(id), id), new JSONDocument(docRoot));
+    }
+    return upsertDocs(upsertMap);
+  }
+
+  private Map<String, String> getDocIdToTenantIdMap(BulkArrayValueUpdateRequest request) {
+    return request.getKeys().stream()
+        .collect(
+            Collectors.toMap(
+                key -> key.toString().split(":")[1], key -> key.toString().split(":")[0]));
+  }
+
+  private BulkUpdateResult bulkAddOnArrayValue(BulkArrayValueUpdateRequest request)
+      throws IOException {
```

Comment:
As discussed:
- can we move common code to the base method?
- can we use the same pattern of reading `existingElems` in all the three methods - prefer `foreach` vs `while`?
- remove `asserts` and assertions issue of docstore test (different result between mongo vs postgres).

Can you file a new issue for `PostgresResultIteratorWithMetaData` for handling `string` columns?